### PR TITLE
stm32: uart: dma circular mode from device tree

### LINKED ISSUE #43308: driver: serial: stm32: uart will lost data when use dma mode[async mode]
**Describe the bug**
User data will be lost when dma buffer is full

e.g.
DMA buffer length is 256 bytes, my package length is 10 bytes, uart driver will lost 4 bytes
when received 26 packages.

Platform is STM32G473QE

### PR DESCRIPTION
**Pull request by fouge**
Allow usage of DMA circular buffering mode from device tree.

The DMA buffer had to be reset when full which lead to missed bytes.
It is now possible to configure the DMA buffer in circular mode by setting the `cyclic` bit.
With the `cyclic` bit, both half and full transfer interrupts are triggered, allowing the user to use the received data while some other bytes are being received in the background in the other half.

To know which interrupt has been triggered, the `status` parameter is used in the DMA callback.
Within the DMA callback, `dma_get_status` is not used by default but the `status` parameter cause by the time the function is called, some new bytes could have been received and we don't want to handle them when processing the interrupt. `UART_RX_BUF_RELEASED` and `UART_RX_BUF_REQUEST` are not used when circular mode is used.

It is still possible for users of the DMA module to force the usage of the circular buffer by setting field `source_reload_en` of `struct dma_block_config` to `1`.

Fixes #43308

### CODE REVIEW

**Comment by fouge** [comment]
@erwango let me know if you need anything more for this PR

File: drivers/dma/dma_stm32.h
Comment (erwango) [inline drivers/dma/dma_stm32.h:29]: Status values are defined by DMA API as [[`include/zephyr/drivers/dma.h`](https://github.com/zephyrproject-rtos/zephyr/blame/dae6d2f84b47cd6c295296dde7dd7d97c40d652d/include/zephyr/drivers/dma.h#LL141)](https://github.com/zephyrproject-rtos/zephyr/blob/dae6d2f84b47cd6c295296dde7dd7d97c40d652d/include/zephyr/drivers/dma.h#L141).

Adding more values would require api change, which is possible but requires discussion and approval from @teburd.
What about using `dma_get_status` existing API ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/dma/dma_stm32.h
Comment (fouge) [reply drivers/dma/dma_stm32.h:29]: Indeed this was my main concern regarding this PR and I don't really like having to use a custom status as well.

As written in the PR description, `dma_get_status` isn't used because the indexes in `struct dma_status` can be modified by the time `uart_stm32_dma_rx_flush` is executed even though the `uart_stm32_dma_rx_flush` is called because the buffer is half or fully filled and the RX buffer should be handled accordingly.

Comment (erwango) [reply drivers/dma/dma_stm32.h:29]: > As written in the PR description,

Right, missed it! txs

Comment (fouge) [reply drivers/dma/dma_stm32.h:29]: What's the status on this? 🤔 
Should we make any positive returned value a custom status that can be used differently depending on the platform?
Microcontrollers have different DMA interrupts so I don't think we want to force the callback to return the same status codes on every MCU.
Should we create an issue for this so that we can discuss it?
cc @teburd

Comment (erwango) [reply drivers/dma/dma_stm32.h:29]: > Should we create an issue for this so that we can discuss it?

This would be the most efficient.

Comment (teburd) [reply drivers/dma/dma_stm32.h:29]: That sounds like a good plan

Comment (fouge) [reply drivers/dma/dma_stm32.h:29]: See #52417 

I actually found that `dma_dw_common.h` stores 2 callbacks in the driver data which might be a good alternative.

**Comment by semihalf-niedzwiecki-dawid** [comment]
Hello :-) what is status of this PR? I would love to use the circular buffer in my application as well.

**Comment by fouge** [comment]
@semihalf-niedzwiecki-dawid Hi! it's currently blocked by the API change needed on the DMA side, see https://github.com/zephyrproject-rtos/zephyr/issues/52417

**Comment by semihalf-niedzwiecki-dawid** [comment]
> @semihalf-niedzwiecki-dawid Hi! it's currently blocked by the API change needed on the DMA side, see #52417

Thanks. Please let me know if you need some help. I can test the PR as well. I'm using the 3Mbps UART with DMA.

File: drivers/flash/flash_stm32_qspi.c
Comment (FRASTM) [inline drivers/flash/flash_stm32_qspi.c:559]: also to be changed for ospi_dma_callback() function in the drivers/flash/flash_stm32_ospi.c  ?  ;-)

Comment (fouge) [reply drivers/flash/flash_stm32_qspi.c:559]: The API change is done in a separate PR #52780

I updated it to include the change in the ospi driver, see https://github.com/zephyrproject-rtos/zephyr/pull/52780/commits/fc05656cb1fba72429513761d80970f0d3476a69

Comment (FRASTM) [reply drivers/flash/flash_stm32_qspi.c:559]: Ok, I was confused  about the PR titlse :  there is more than use of dma circular buffer for the uart

File: drivers/dma/dma_stm32.c
Comment (marcowidmer) [inline drivers/dma/dma_stm32.c:126]: This currently breaks transmitting with spi_ll_stm32. When calling spi_write() consecutively, I get the following error:
```
<err> dma_stm32: dma stream 1 is busy.
<err> dmamux_stm32: cannot configure the dmamux.
```
stream->busy is not cleared in `dma_stm32_irq_handler` because the direction is `MEMORY_TO_PERIPHERAL` and it is also not cleared in `dma_stm32_stop` because `stm32_dma_is_enabled_stream()` returns false (the hardware channel has already stopped).

Please remove `stream->direction == PERIPHERAL_TO_MEMORY`

Comment (marcowidmer) [reply drivers/dma/dma_stm32.c:126]: Btw, shouldn't we also check `config->head_block->source_reload_en` here?

**Comment by fouge** [comment]
I cannot re-open
I am going to wait for the DMA API change to take effect and will likely submit another PR.

I took @marcowidmer comments into account and I actually want to provide tests cause we had issues with using this driver in low power.

**Pull request closed without merging**
